src/TermInterface.jl

```diff
@@ -1,14 +1,36 @@
+"""
+This module defines a contains definitions for common functions that are useful
+for symbolic expression manipulation. Its purpose is to provide a shared
+interface between various symbolic programming Julia packages.
+
+This is currently borrowed from TermInterface.jl. If you want to use
+Metatheory.jl, please use this internal interface, as we are waiting that a
+redesign proposal of the interface package will reach consensus. When this
+happens, this module will be moved back into a separate package.
+
+See https://github.com/JuliaSymbolics/TermInterface.jl/pull/22
+"""
 module TermInterface
 
 """
   istree(x)
 
-Returns `true` if `x` is a term. If true, `operation`, `arguments`
-must also be defined for `x` appropriately.
+Returns `true` if `x` is a term. If true, `head`, `children` and 
+`is_function_call` must also be defined for `x` appropriately.
 """
 istree(x) = false
 export istree
 
+"""
+  is_function_call(x)
+
+Returns true if a term abstractly represents a function call or function application.
+Must be defined if `istree(x)` is defined. 
+Can be true only if `istree(x)` is true.
+"""
+function is_function_call end
```

Comment:
Why not default this to false? If it's an optional interface, why do we require users to define it?

Reply:
I would not make this optional. Defaulting it to false may cause users to forget to set it on specific term types, and thus cause hard to catch bugs as it would propagate through all the rewriting steps. 

In comparison to the previous state of this redesign, where we had `Head` types, in this latest version, the provider of a term types `T <: AbstractT` just has to set `is_function_call(<:AbstractT) = true` if the whole language is functional, or set it to false if the entire AST does not have function calls